File: WebClient SDK/src/PnrScripts/UI/PnrOrderUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { Guid, createTypedField, DotNetType, Visibility } from 'tessa/platform';
import { CardStoreMode } from 'tessa/cards';
import { PnrCardTypes } from '../Shared/PnrCardTypes';
import { PnrRequestTypes } from '../Shared/PnrRequestTypes';
import { PnrDocumentKinds } from '../Shared/PnrDocumentKinds';
import { CardRequest, CardService } from 'tessa/cards/service';
import { DateTimeViewModel } from 'tessa/ui/cards/controls';
import Moment from 'moment';

export class PnrOrderUIExtension extends CardUIExtension {
  public async initialized(context: ICardUIExtensionContext) {
    if (!Guid.equals(context.card.typeId, PnrCardTypes.PnrOrderTypeID)) {
      return;
    }
    let PnrOrder = context.card.sections.tryGet("PnrOrder");
    if (!PnrOrder) return;
    // Создание карточки
    if (context.card.storeMode == CardStoreMode.Insert)
    {
      // Заполнить поле Дата регистрации текущей датой
      const regDateControl = context.model.controls.get('RegistrationDate') as DateTimeViewModel;
      if (regDateControl) {
        regDateControl.selectedDate = Moment(Date.now());
      }
      let dci = context.card.sections.tryGet("DocumentCommonInfo");
      if (dci != null) {
        // начальная инициализация поля Подразделение по автору
        await this.SetUserDepartmentInfo(dci.fields.get('AuthorID'), context);
      }
    }
    else { // Открытие карточки. Есть Вид входящего документа - устанавливаем видимость контролов в зависимости от Вида
      let documentKindID = PnrOrder.fields.get('DocumentKindID');
      if (documentKindID != null) {
        this.SetDKDependenceVisibility(documentKindID, context);
      }
    }

    PnrOrder.fields.fieldChanged.add(async (e) => {
      // Вид документа: Visibility контролов и переформирование номера
      if (e.fieldName == "DocumentKindID" && e.fieldValue != null)
      {
         this.SetDKDependenceVisibility(e.fieldValue, context);
      }
    });
  }

  private async SetUserDepartmentInfo(authorID, context: ICardUIExtensionContext)
  {
    let DepartmentID;
    let DepartmentName;
    let DepartmentIdx;
    let request = new CardRequest();
        request.requestType = PnrRequestTypes.GetUserDepartmentInfoRequestTypeID;
        request.info["authorID"] = createTypedField(authorID, DotNetType.Guid);

    // проверка ответа
    let response = await CardService.instance.request(request);
    if (response.validationResult.isSuccessful) {
      DepartmentID = response.info.DepartmentID;
      DepartmentName = response.info.Name;
      DepartmentIdx = response.info.Index;
    }

    let PnrOrder = context.card.sections.tryGet("PnrOrder");
    if (DepartmentID != null && DepartmentName != null && DepartmentIdx != null && PnrOrder != null) {
      PnrOrder.fields.set('DepartmentID', createTypedField(DepartmentID.$value, DotNetType.Guid));
      PnrOrder.fields.set('DepartmentName', createTypedField(DepartmentName.$value, DotNetType.String));
      PnrOrder.fields.set('DepartmentIdx', createTypedField(DepartmentIdx.$value, DotNetType.String));
    }
  }

  /** установка видимости контролов, зависимых от Вида документа */
  private SetDKDependenceVisibility(documentKindID, context: ICardUIExtensionContext)
  {
    const disposalBlock = context.model.blocks.get('DisposalBlock');
    if (!disposalBlock) return;
    if (Guid.equals(documentKindID, PnrDocumentKinds.OrderAdministrativeActivity) ||
        Guid.equals(documentKindID, PnrDocumentKinds.OrderMainActivity) ||
        Guid.equals(documentKindID, PnrDocumentKinds.OrderMobileCommunications) ||
        Guid.equals(documentKindID, PnrDocumentKinds.OrderImplementation))
    {
      disposalBlock.blockVisibility = Visibility.Collapsed;
    }
    else if (Guid.equals(documentKindID, PnrDocumentKinds.Disposal))
    {
      disposalBlock.blockVisibility = Visibility.Visible;
    }
  }
}

File: Extensions/Tessa.Extensions.Client/UI/PnrOrderUIExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Shared.PnrCards;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Client.UI
{
    public sealed class PnrOrderUIExtension : CardUIExtension
    {
        private readonly ICardRepository cardRepository;
        public PnrOrderUIExtension(ICardRepository cardRepository)
        {
            this.cardRepository = cardRepository;
        }

        private async void SetUserDepartmentInfo(Guid? authorID, ICardModel cardModel)
        {
            CardRequest request = new CardRequest
            {
                RequestType = Shared.PnrRequestTypes.GetUserDepartmentInfoRequestTypeID,
                Info =
                {
                    { "authorID", authorID }
                }
            };

            CardResponse response = await cardRepository.RequestAsync(request);

            Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
            TessaDialog.ShowNotEmpty(result);
            if (result.IsSuccessful)
            {
                cardModel.Card.Sections["PnrOrder"].Fields["DepartmentID"] = response.Info.Get<Guid?>("DepartmentID");
                cardModel.Card.Sections["PnrOrder"].Fields["DepartmentName"] = response.Info.Get<string>("Name");
                cardModel.Card.Sections["PnrOrder"].Fields["DepartmentIdx"] = response.Info.Get<string>("Index");
            }
        }

        // установка видимости контролов, зависимых от Вида документа
        private void SetDKDependenceVisibility(Guid? documentKindID, ICardModel cardModel)
        {
            //IBlockViewModel orderAdministrativeActivityBlock = cardModel.Blocks["OrderAdministrativeActivityBlock"];
            //orderAdministrativeActivityBlock.BlockVisibility = documentKindID == PnrDocumentKinds.OrderAdministrativeActivity ? Visibility.Visible : Visibility.Collapsed;

            //IBlockViewModel orderMainActivityBlock = cardModel.Blocks["OrderMainActivityBlock"];
            //orderMainActivityBlock.BlockVisibility = documentKindID == PnrDocumentKinds.OrderMainActivity ? Visibility.Visible : Visibility.Collapsed;

            //IBlockViewModel orderMobileCommunicationsBlock = cardModel.Blocks["OrderMobileCommunicationsBlock"];
            //orderMobileCommunicationsBlock.BlockVisibility = documentKindID == PnrDocumentKinds.OrderMobileCommunications ? Visibility.Visible : Visibility.Collapsed;

            //IBlockViewModel orderImplementationBlock = cardModel.Blocks["OrderImplementationBlock"];
            //orderImplementationBlock.BlockVisibility = documentKindID == PnrDocumentKinds.OrderImplementation ? Visibility.Visible : Visibility.Collapsed;

            //IBlockViewModel disposalBlock = cardModel.Blocks["DisposalBlock"];
            //disposalBlock.BlockVisibility = documentKindID == PnrDocumentKinds.Disposal ? Visibility.Visible : Visibility.Collapsed;

            //disposalBlock.Rearrange();

            IBlockViewModel disposalBlock = cardModel.Blocks["DisposalBlock"];
            if (documentKindID == PnrDocumentKinds.OrderAdministrativeActivity ||
                documentKindID == PnrDocumentKinds.OrderMainActivity ||
                documentKindID == PnrDocumentKinds.OrderMobileCommunications ||
                documentKindID == PnrDocumentKinds.OrderImplementation)
            {
                disposalBlock.BlockVisibility = Visibility.Collapsed;
            }
            else if (documentKindID == PnrDocumentKinds.Disposal)
            {
                disposalBlock.BlockVisibility = Visibility.Visible;
            }
            disposalBlock.Rearrange();

        }

        public override Task Initialized(ICardUIExtensionContext context)
        {
            Card card = context.Card;
            ICardModel cardModel = context.Model;

            // Создание карточки
            if (card.StoreMode == CardStoreMode.Insert)
            {
                // Заполнить поле Дата регистрации текущей датой
                if (cardModel.Controls.TryGet("RegistrationDate", out IControlViewModel regDateControl))
                {
                    ((DateTimeViewModel)regDateControl).SelectedDate = DateTime.Now;
                }

                // начальная инициализация поля Подразделение по автору
                SetUserDepartmentInfo((Guid?)card.Sections["DocumentCommonInfo"].Fields["AuthorID"], cardModel);
            }
            // Открытие карточки. Есть Вид входящего документа - устанавливаем видимость контролов в зависимости от Вида
            else
            {
                Guid? documentKindID = (Guid?)card.Sections["PnrOrder"].Fields["DocumentKindID"];
                if (documentKindID != null)
                    SetDKDependenceVisibility(documentKindID, cardModel);
            }

            context.Card.Sections["PnrOrder"].FieldChanged += (s, e) =>
            {
                // Вид документа: Visibility контролов и переформирование номера
                if (e.FieldName == "DocumentKindID" && e.FieldValue != null)
                {
                    SetDKDependenceVisibility((Guid)e.FieldValue, cardModel);
                }
            };

            return base.Initialized(context);
        }
    }
}
